lib/aur-chroot

```diff
@@ -184,88 +210,147 @@ elif [[ ! -f $makepkg_conf ]]; then
-    fi
+    # # create both the root container and a directory for pacman db
+    install -d "$directory"/root/var/lib/pacman -m 755
+
+    # # XXX: allow setting machine-id from host (and provide default) to enable .nspawn file usage
+    systemd-firstboot \
+        --root="$directory"/root \
+        --copy-locale \
+        --copy-timezone \
+        --setup-machine-id
+
+    pacman -Sy "${base_packages[@]}" \
+        --config "$pacman_conf" \
+        -r "$directory"/root \
+        "${cache_dirs[@]/#/--cachedir=}" \
+        --noconfirm
+
+    {
+        declare -f _chroot_create
+        printf '_chroot_create || exit\n'
+    } >> "$directory"/root/aur-chroot--helper
 fi >&2
 
-# arch-nspawn makes no distinction between a missing working directory
-# and one which does not exist
-if [[ ! -d $directory/root ]]; then
-    printf >&2 '%s: %q is not a directory\n' "$argv0" "$directory"/root
-    printf >&2 '%s: did you run aur chroot --create?\n' "$argv0"
-    exit 20
-fi
-
 if (( update )); then
-    # locking is done by systemd-nspawn
-    # shellcheck disable=SC2086
-    $AUR_PACMAN_AUTH arch-nspawn -C "$pacman_conf" -M "$makepkg_conf" "$directory"/root \
-         "${bindmounts_ro[@]/#/--bind-ro=}" \
-         "${bindmounts_rw[@]/#/--bind=}" pacman -Syu --noconfirm "$@"
+    printf 'pacman -Syu --noconfirm --config %q %s || exit' "$pacman_conf" "${*@Q}" >> "$directory"/root/aur-chroot--helper
 fi >&2
 
-# print path for processing by other tools (e.g. makepkg --packagelist)
-if (( print_path )); then
-    realpath -- "$directory"/root
-    exit $?
-fi
+if (( create )) || (( update )); then
+    systemd-nspawn \
+        -D "$directory"/root \
+        "${bindmounts_ro[@]/#/--bind-ro=}" \
+        "${bindmounts_rw[@]/#/--bind=}" \
+        /bin/bash /aur-chroot--helper
+
+    exit
+fi >&2
 
 if (( build )); then
-    # use makechrootpkg -c as default build command (sync /root container)
-    # arguments after -- are used as makechrootpkg arguments
-    # shellcheck disable=SC2086
-    $AUR_PACMAN_AUTH "${auth_args[@]}" makechrootpkg -r "$directory" \
-         "${bindmounts_ro[@]/#/-D}" "${bindmounts_rw[@]/#/-d}" \
-         "${makechrootpkg_args[@]}" -- "${makechrootpkg_makepkg_args[@]}"
+    # add configurable directories for makepkg (and PWD as BUILDDIR)
+    bindmounts_rw+=(
+        "$PWD"
+        "${PKGDEST:-$PWD}:/pkgdest"
+        "${LOGDEST:-$PWD}:/logdest"
+        "${SRCDEST:-$PWD}:/srcdest"
+        "${SRCPKGDEST:-$PWD}:/srcpkgdest"
+    )
+
+    # create makepkg_user
+    systemd-sysusers --root "$directory"/root --replace=/usr/lib/sysusers.d/build.conf - \
+        < <(printf "u %q %s \"makepkg user\" /build /bin/bash\n" "$makepkg_user" "$AUR_CHROOT_UID")
+
+    # allow makepkg_user to install packages (e.g. --syncdeps)
+    tee "$directory/root/etc/sudoers.d/makepkg_user" \
+        < <(printf '%q ALL = NOPASSWD: /usr/bin/pacman' "$makepkg_user") \
```

Comment:
`%q` is a form of encoding that produces strings according to bash rather than sudo. The `User` grammar for sudo doesn't allow single quotes but uses similar mechanism for escaping problem characters:

Consider the `sudoers` manual:

```
User_List ::= User |
              User ',' User_List

User ::= '!'* user name |
         '!'* #user-ID |
         '!'* %group |
         '!'* %#group-ID |
         '!'* +netgroup |
         '!'* %:nonunix_group |
         '!'* %:#nonunix_gid |
         '!'* User_Alias

[...]

A user  name,  user-ID,  group,  group-ID,  netgroup,  nonunix_group  or
nonunix_gid  may  be enclosed in double quotes to avoid the need for es‐
caping special characters.  Alternately, special characters may be spec‐
ified in escaped hex mode, e.g., \x20  for  space.   When  using  double
quotes, any prefix characters must be included inside the quotes.

[...]

Quotes around group names are optional.  Unquoted  strings  must  use  a
backslash  (‘\’)  to  escape  spaces and special characters.  See “Other
special characters and reserved words” for a  list  of  characters  that
need to be escaped.
```

Here quotes are `"` which `%q` won't use, but it might use `\`. So I would be careful about mixing encoding schemes and making sure they accord to one another.

(N.B. sudoers doesn't seem to say how to handle `"` inside a quoted string, whether `\"` or closing, escaping and re-opening is needed (`"\""`))

Reply:
thanks for taking the time to explain this!